lib/shipping_api.rb

```diff
@@ -0,0 +1,127 @@
+module FedAxApiWrapper
+  class ShippingApi
+    DESIRED_USPS_SERVICES = [
+      "USPS First-Class Mail Parcel",
+      "USPS Standard Post",
+      "USPS Priority Mail 2-Day Medium Flat Rate Box",
+      "USPS Priority Mail 2-Day Large Flat Rate Box",
+      "USPS Priority Mail Express 2-Day Flat Rate Boxes"
+    ]
+
+    def self.query(delivery_info)
+      # delivery_info is the packages, origin, and destination information passed
+      # into the API from the user -- in this case, bEtsy.
+      ups_response = self.ups_query(delivery_info)
+      ups_quotes = self.extract_data_from_ups_response(ups_response)
+
+      usps_response = self.usps_query(delivery_info)
+      usps_quotes = self.extract_data_from_usps_response(usps_response)
+
+      quotes = {}
+      quotes[:ups] = ups_quotes
+      quotes[:usps] = usps_quotes
+
+      if quotes[:ups].empty? || quotes[:usps].empty?
+        response = { status: 204 }
+      else
+        response = { quotes: quotes, status: 200 }
+      end
+    end
+
+    def self.extract_data_from_ups_response(response)
+      combined_results = []
+
+      # this is made available to us by the ActiveShipping gem.
+      rates = response.rates
+
+      rates.each do |rate|
+        result = {}
+
+        # these methods are also made available to us by the ActiveShipping gem.
+        result["carrier"] = rate.carrier # "UPS"
+        result["total_price"] = rate.total_price # price in cents
+        result["service_type"] = rate.service_name # "UPS Ground"
+        result["expected_delivery"] = rate.delivery_date # nil or timedate
+
+        combined_results.push(result)
+      end
+
+      return combined_results
+    end
+
+    def self.extract_data_from_usps_response(response)
+      combined_results = []
+
+      # this method is made available to us by the ActiveShipping gem.
+      rates = response.rates
+
+      rates.each do |rate|
+        if DESIRED_USPS_SERVICES.include? rate.service_name
+          result = {}
+
+          # these methods are also made available to us by the ActiveShipping gem.
+          result["carrier"] = rate.carrier # "USPS"
+          result["total_price"] = rate.total_price # price in cents
+          result["service_type"] = rate.service_name # "USPS Standard Post"
+          result["expected_delivery"] = rate.delivery_date # nil or timedate
+
+          combined_results.push(result)
+        end
+      end
+
+      return combined_results
+    end
```

Comment:
Great job with these. I like how you extracted out some methods and made them more concise. It makes the query method easy to follow which is great!!!

One thing to note is that the self.extract_data_from_ups_response(response) method and the self.extract_data_from_usps_response(response) method look almost exactly the same. It might be interesting to try to make a general method that would cover both of them. (Looks like the really tricky part would be line 59)